### ISSUE TITLE
Automatic Resource Detection in dotnet similar to Java

### ISSUE THREAD

**Issue by swetharavichandrancisco** [open-telemetry/opentelemetry-dotnet#2752]
Purpose of the Feature Request:
-----------------------------------
Creation of Automatic Resource Attributes, especially looking to add container.id to resource metadata
Corresponding JAVA PR posted below, 
- https://github.com/open-telemetry/opentelemetry-java/pull/3321

Current Behavior in Java:
--------------------------
In Java, We have ResourceProvider Interface that are implemented in the below classes 
- https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/resources/src/main/java/io/opentelemetry/sdk/extension/resources

All these Providers are read OOTB in OpenTelemetryResourceAutoConfiguration
- https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java#60

and it is configured in the below file,

- https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/aws/src/main/resources/META-INF/services/io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider

Current Behavior in DotNet:
------------------------------
In DotNet, We have ResourceBuilderExtensions class that adds Semantic Attributes with Dedicated Environment Variable/SDK-provided Default Value like service, opentelemetry SDK
- https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Resources/ResourceBuilderExtensions.cs

From the customer application, we still have to call the respective api's for the resources to be added.
Below is the sample code,
`var tracerProvider = Sdk.CreateTracerProviderBuilder()
                        .SetResourceBuilder(ResourceBuilder.CreateEmpty().AddService("test"))
                        .AddSource("Sample.DistributedTracing")
                        .AddConsoleExporter()
                        .Build()`
Currently, there is no way to add Resources automatically.

Expected/Proposed Solution:
--------------------------------
Can we have something similar to Java? Implement ResourceProviders in DotNet and ResourceBuilder class can pick all the ResourceProviders. We should also provide a way to configure ResourceProviders.

Thanks.

**Comment by swetharavichandrancisco** [comment]
I can work on the feature as well

**Comment by alanwest** [comment]
> I can work on the feature as well

@swetharavichandrancisco sounds good!

The specification has a few words about resource detectors that my prove useful https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#detecting-resource-information-from-the-environment

**Comment by swetharavichandrancisco** [comment]
Thank you @alanwest

**Comment by alanwest** [comment]
@swetharavichandrancisco Yes, your contributions would be welcome! I think it would be great to additional resource detectors for things like Docker.

I think the title of that section of the spec is misleading... it is not just about detection via environment variables, but more generally the environment the application is running in like Docker or a cloud vendor. It provides some guidance that you'll like need to take into account.

> Custom resource detectors related to generic platforms (e.g. Docker, Kubernetes) or vendor specific environments (e.g. EKS, AKS, GKE) MUST be implemented as packages separate from the SDK.

So, just as Java has separated out their resource detectors into a package, we'll likely need to do the same.

If you're proposing to build both additional resource detectors and also an auto resource configuration mechanism like Java, I think it would be best to split up these efforts.

**Comment by cijothomas** [comment]
Exposing this as public is likely all that we need : https://github.com/open-telemetry/opentelemetry-dotnet/blob/6b7f2dd77cf9d37260a853fcc95f7b77e296065d/src/OpenTelemetry/Resources/IResourceDetector.cs#L22
